### ISSUE TITLE
Why is it no longer available to return chunks in the 'get' method?

### ISSUE THREAD

**Issue by SteveFlames** [theophilusx/ssh2-sftp-client#314]
Hello! In version 3 of this package, the 'get' method used to return chunks.
I was wondering why is it no longer available?

This was handy in case we don't want to write the file locally but instead use it, for example, to upload it straight to S3.

**Comment by theophilusx** [comment]
To be honest, I don't recall precisely what the changes were - version 3
was some time ago. From memory, I do recall that some of the buffer
handling options had to be changed because of changes in the buffer API.
Any API changes were made to increase reliability with a Promise based
wrapper around an event based library (ssh2). I don't believe any
functionality has been lost as anything you could do before you can do
with either a buffer or a stream. 

Note that you don't have to write the file locally. You can supply a
stream object to the get method and that stream object can do anything
you want, including breaking the data into 'chunks' of data. In fact,
you can create whole pipelines of stream objects that perform all sorts
of data transformation. You can define an s3 stream object which takes
the data, transforms it in any way you want, possibly breaking it into
specific chunk sizes and then write it to s3. 

There are also numerous s3 stream packages available on npmjs.com. Any
package which provides a stream object could be used as the target for
the get method. You can use one of these packages to upload the data
directly from the get method call instead of creating your own stream object.

**Comment by SteveFlames** [comment]
Thanks for the quick response!

The truth is I tried using a newer version and providing a stream to the 'get' method as you suggested, but I am having problems with big files; it hangs.

I managed to reproduce it in a simple example, so here it is:
Using `node v12.22.1` and `ssh2-sftp-client v5.3.2`
```
const SFTPClient = require('ssh2-sftp-client')
const { PassThrough, Readable } = require('stream')

async function main () {
  const remoteFilepath = './file.js'

  const nOfCharacters = 100000 // With 10000 it does not hang
  const file = generateBigFileAsStream(nOfCharacters)

  await putFile(remoteFilepath, file)
  const stream = await readFile(remoteFilepath)

  return new Promise((resolve, reject) => {
    stream.pipe(process.stdout)
    stream.on('end', () => resolve())
    stream.on('error', (err) => reject(err))
  })
}

function generateBigFileAsStream (n) {
  let largeBuffer = Buffer.from('')
  const anotherBuffer = Buffer.from('a')

  for (let i = 0; i < n; i++) {
    largeBuffer = Buffer.concat([largeBuffer, anotherBuffer])
  }

  const file = new Readable()
  file.push(largeBuffer)
  file.push(null)
  return file
}

async function readFile (filepath) {
  const sftp = await getConnection()
  console.log('Reading file through sftp', filepath)
  try {
    const encoding = null

    const stream = await sftp.get(filepath, new PassThrough(), { encoding }) // Hangs here

    stream.on('end', function () {
      console.log('Closing sftp connection since stream finished')
      sftp.end()
    })
    stream.on('error', function (err) {
      console.error('Error with sftp stream', err)
      sftp.end()
      stream.destroy()
    })
    return stream
  } catch (e) {
    console.error('Error reading file as stream, closing now connection. Error:', e)
    sftp.end()
    throw e
  }
}

async function putFile (filepath, stream) {
  const sftp = await getConnection()
  try {
    await sftp.put(stream, filepath)
  } catch (e) {
    console.error('Error writing file in sftp', filepath)
    throw e
  } finally {
    sftp.end()
  }
}

async function getConnection () {
  try {
    const sftp = new SFTPClient()
    // TODO: fill in the sftp creds
    const sftpParams = {
      host: sftpConfig.host,
      port: sftpConfig.port,
      username: sftpConfig.username,
      password: sftpConfig.password
    }
    console.log('Connecting to sftp')
    await sftp.connect(sftpParams)
    console.log('Connected to sftp')
    return sftp
  } catch (e) {
    console.error('Error getting connection')
    console.error(e.message)
    const merr = new Error('SFTP connection problem')
    throw merr
  }
}

main()
  .then(() => {
    console.log('All done!')
  })
  .catch((err) => {
    console.error('Error in main', err)
  })
```

I am aware that this might be a totally different issue than the one I opened, but this is the reason I was using v3.1.0, because it was working for me without a problem for big files also.

The thing is when I upgraded from node v8 to node v12 then even v3.1.0 started hanging.

Am I missing something in the code above?
I would expect it to work the same regardless the file size.

**Comment by theophilusx** [comment]
Please use ssh2-sftp-cleint v6.0.1 

I think there are a number of errors in your code. I should be able to
find some time today to knock up a very simple example you will be able
to build on. There are quite a lot of people using the module, including
myself, to download large files. I use a gzip/gunzip stream to
compress/decompress data 'on the fly' and regularly download multi-Gb
sized files with no problems. Others have reported successfully using
streams to implement download/upload transfer rate calculations,
implement rate limiting etc. The only data size issue we have
encountered has been a problem with one brand of sftp server
(download.com) which causes problems with fastGet/fastPut and a problem
with node v14.x due to a change in the network library which was rolled
back in node v15.3. 

One of the big challenges when trying to use a library which is event
based (ssh2), but with a Promise based API (ssh2-sftp-cleint) is
handling events correctly. The pipe() method significantly simplifies
this process. In 99% of use cases for ssh2-sftp-client, there is no need
to add any 'on' listeners. The pipe() process will automatically handle
the closing of streams and event handlers built into the methods will
handle errors etc. Nearly all the changes within the module with the
last few versions have been about refinement of the internal model to
handle events (for example, removing handlers to prevent memory leaks
and managing possible events raised 'outside' or 'in between' Promises. 

Working with streams can be a little challenging initially. It is
actually very simple. However, it can seem confusing because the stream
library is extremely flexible and provides the ability to control almost
any aspect of the stream. The reality is that in most cases, you don't
need that level of power. Have a read of the npm 'stream-handbook'
package. It is slightly outdated now, but not a bad starting point for
working with streams.

I've made some very quick hacks to your below code. I've not tested
this, but it should be close to what would work. I'll try and knock up
something simple later today.

Note that below, I have assumed there is already a test file on the
remote sftp server for download. You can use something like openSSH's
sftp CLI to copy a test file up to the remote sftp server. This allows
us to just focus on the getting and processing of the data without other
noise.

> const SFTPClient = require('ssh2-sftp-client')
> const { PassThrough, Readable } = require('stream')
>
> async function main () {
>   const remoteFilepath = './file.js'
>
>   await readFile(remoteFilepath)
>
>   //You don't want to do this. 
>   //return new Promise((resolve, reject) => {
>   //  stream.pipe(process.stdout)
>   //  stream.on('end', () => resolve())
>   //  stream.on('error', (err) => reject(err))
>   //})
> }
>
>
> async function readFile (filepath) {
>   const sftp = await getConnection()
>   console.log('Reading file through sftp', filepath)
>   try {
>     const encoding = null
>     const pt = new PassThrough();
>     const target = pt.pipe(process.stdout);
>     await sftp.get(filepath, target, { encoding });
>     console.log('Transfer complete');
>     await sftp.end();
>   } catch (e) {
>     console.error('Error reading file as stream, closing now connection. Error:', e)
>     sftp.end()
>     throw e
>   }
> }
>
> async function getConnection () {
>   try {
>     const sftp = new SFTPClient()
>     // TODO: fill in the sftp creds
>     const sftpParams = {
>       host: sftpConfig.host,
>       port: sftpConfig.port,
>       username: sftpConfig.username,
>       password: sftpConfig.password
>     }
>     console.log('Connecting to sftp')
>     await sftp.connect(sftpParams)
>     console.log('Connected to sftp')
>     return sftp
>   } catch (e) {
>     console.error('Error getting connection')
>     console.error(e.message)
>     const merr = new Error('SFTP connection problem')
>     throw merr
>   }
> }
>
> main()
>   .then(() => {
>     console.log('All done!')
>   })
>   .catch((err) => {
>     console.error('Error in main', err)
>   })
>

-- 
Tim Cross

**Comment by SteveFlames** [comment]
I see.. I was relying more on events when I shouldn't, since this is a promise based library.

Nevertheless, I still didn't get it to work. Even with your proposal it hangs.
I am wondering if it could be the sftp server's fault.

In any case, thanks for the recommendations! I'll keep trying to debug this

**Comment by theophilusx** [comment]
OK, found time to look at this.

The hanging you have observed is due to piping to stdout and the Promise
waiting until that stream is 'finished'. Problem is, stdout never
'finishes' - at least not until the node process exits.

If you use a different stream, one which does 'finish', like a file
write stream, all works fine. Also works with other streams which can be
closed. However, I can also see a use case for writing to a destination
stream which you don't want closed/finished at the end of the piping of
data.

I'll make some changes and let you know when I've completed them. You
can then try out the code in the repo and see if that now works for you.

**Comment by theophilusx** [comment]
I have made a small change to the get() method which makes it work with
a destination stream which doesn't finish/close. This should resolve the
hang issue you observed when trying to pipe to stdout.

I also added a very simple script which retrieves a file, pipes it
through PassThrough to process.sdtout. See examples/get-stream.js in the
git repository.

Try using the repository version and see if that helps. If it does, I'll
release a new minor version.

**Comment by SteveFlames** [comment]
This works indeed in the simple example, when doing it with the flow you suggested (first pipe all the streams and then call the get() method last)

My only concern is that this flow feels a bit weird for me. I would expect to first call the get() method to retrieve the stream and then pass it on and use it wherever I want.
This was the case with v3 of the library as the get() method resolved as soon as the "readable" event was fired, returning the readable stream.

On the contrary, in newer versions I see that get() resolves on "finish" (or "end" with this current fix) so I guess it always has to be called last.

It just seems a bit unintuitive to me! Maybe that could fit in a new function like getStream()? I see that was mentioned also [here](https://github.com/theophilusx/ssh2-sftp-client/issues/90).

Wdyt?

**Comment by theophilusx** [comment]
This change was made to make the API more consistent as a Promise based
API. There were quite a lot of problems with the 3.x version (number of
bugs and event handling was severely broken). There are a number of
challenges when trying to wrap an event based API (ssh2) in a Promise
based API. If just the stream is returned, there is no way of knowing
when use of the stream has been completed. This means that the
client code would need to manage all event handlers and making
sure they are removed to prevent memory leaks once no longer needed. 

With the current API, everything is wrapped up in the Promise. Once the
promise has been fulfilled, you know that all resources and event
listeners have been released and the operation has either succeeded
(resolved) or failed (rejected). The code using the API doesn't need to
worry about adding/removing event listeners. 

The way to conceptualise this model is that you first create the
pipeline (or even 'bucket') where you want the data to be put and then
call the get method, passing in the pipeline/bucket where you want the
data to go.

If you want an API which gives you greater control, such as passing back
the readStream and then allowing you to do whatever you want with it,
you also need to manage the event handlers. If you are going to do that,
you may as well just use the ssh2 API and skip ssh2-sftp-client.

**Comment by SteveFlames** [comment]
You are totally right :ok_hand: 

I will give it another try to use the API with promises as intended.

Thanks again for the valuable and descriptive info!
